src/librustc_typeck/collect/type_of.rs

```diff
@@ -177,13 +177,8 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
 
         Node::Field(field) => icx.to_ty(&field.ty),
 
-        Node::Expr(&Expr { kind: ExprKind::Closure(.., gen), .. }) => {
-            let substs = InternalSubsts::identity_for_item(tcx, def_id);
-            if let Some(movability) = gen {
-                tcx.mk_generator(def_id, substs, movability)
-            } else {
-                tcx.mk_closure(def_id, substs)
-            }
+        Node::Expr(&Expr { kind: ExprKind::Closure(..), .. }) => {
+            tcx.typeck_tables_of(def_id.expect_local()).node_type(hir_id)
```

Comment:
This feels wrong to me. We basically *undid* a move much like this for generators, though partly it was to avoid those type error problems you were referring to with your other change. 

But also it is just surprising to me to have `type_of` require us to type-check the closure. For every other def-id, getting `type_of` doesn't require a full type-check, but rather returns a "fully generic" type. i.e., I would expet the type of a closure or generator ought to be expressed as a purely generic type like `Closure<P0...Pn>`, much like we'd expect for a function, not one that embeds full types like `Closure<..., (u32, u64)>` or whatever. Do we have any other example that works like that?

Anyway, what is the exact motivation here? It's confusing to me :)

Reply:
Ah, ok, I see, the point is that you want `Instance` to be able to substitute *only* the `base_params` and to still get a valid type... hmm. Interesting. I'm not sure what I think about it. I'm surprised this gives much of a win, can you explain why it does?

Reply:
If we do land this, I feel like we need some kind of comment..somewhere...that explains this rather subtle point. I guess it would be best to document it on the `ClosureSubsts` and `GeneratorSubsts` structs.

Reply:
> Ah, ok, I see, the point is that you want `Instance` to be able to substitute only the `base_params` and to still get a valid type... hmm. Interesting. I'm not sure what I think about it.

Me neither, I think I'd prefer if we could implement e.g. `fn_sig` on closures instead, and use that from `Instance`.
I could look into that approach once #74314 lands.

> I'm surprised this gives much of a win, can you explain why it does?

All the wins are incremental, so it probably only makes sense if a closure changes its capture set/types, although that would still imply codegen is needed?

Or maybe non-generic closures are involved, which would now result in an `Instance` with empty substs, which may be handled differently by CGUs.